compiler/rustc_metadata/src/rmeta/decoder/cstore_impl.rs

```diff
@@ -311,54 +312,74 @@ pub fn provide(providers: &mut Providers) {
-            {
-                let visible_parent_map = &mut visible_parent_map;
-                let mut add_child =
-                    |bfs_queue: &mut VecDeque<_>, child: &Export<hir::HirId>, parent: DefId| {
-                        if child.vis != ty::Visibility::Public {
-                            return;
-                        }
-
-                        if let Some(child) = child.res.opt_def_id() {
-                            match visible_parent_map.entry(child) {
-                                Entry::Occupied(mut entry) => {
-                                    // If `child` is defined in crate `cnum`, ensure
-                                    // that it is mapped to a parent in `cnum`.
-                                    if child.is_local() && entry.get().is_local() {
-                                        entry.insert(parent);
-                                    }
-                                }
-                                Entry::Vacant(entry) => {
-                                    entry.insert(parent);
-                                    bfs_queue.push_back(child);
-                                }
-                            }
-                        }
-                    };
+            // Iterate over parents vector to remove duplicate elements
+            // while preserving order
+            let mut dedup_set = FxHashSet::default();
+            for (_, parents) in &mut visible_parents_map {
+                parents.retain(|parent| dedup_set.insert(*parent));
 
-                while let Some(def) = bfs_queue.pop_front() {
-                    for child in tcx.item_children(def).iter() {
-                        add_child(bfs_queue, child, def);
-                    }
-                }
+                // Reuse hashset allocation.
+                dedup_set.clear();
             }
 
-            visible_parent_map
+            Lrc::new(visible_parents_map)
         },
+        best_visible_parent: |tcx, child| {
+            // Use `min_by_key` because it returns
+            // first match in case keys are equal
+            tcx.visible_parents_map(())
+                .get(&child)?
+                .into_iter()
+                .min_by_key(|parent| {
+                    // If this is just regular export in another module, assign it a neutral score.
+                    let mut score = 0;
+
+                    // If child and parent are local, we prefer them
+                    if child.is_local() && parent.is_local() {
+                        score += 1;
+                    }
+
+                    // Even if child and parent are local, if parent is `#[doc(hidden)]`
+                    // We should prefer it only as last resolve
+                    // If other items are not available
+                    if tcx.is_doc_hidden(**parent) {
+                        score -= 2;
+                    }
 
+                    -score
```

Comment:
You can avoid negating this by using `max_by_key` instead.

Reply:
@jyn514 this will select last item, instead of first

Reply:
https://doc.rust-lang.org/stable/core/iter/trait.Iterator.html#method.max_by_key
> If several elements are equally maximum, the **last** element is returned. If the iterator is empty, None is returned.

Reply:
Ok - that's not evident just from the code, can you leave a comment?

Also, why does min_by_key behave differently? I don't understand why you would prefer the first over the last or vice versa.

Reply:
@jyn514 Preferring first was just what algorithm did before this PR.
Preferring last will produce slightly results and I think will prevent early exit from loop.

For example replacing `min_by_key` by `max_by_key` will produce following failure in one of unit tests.

```rust
---- [ui] ui/trait-impl-bound-suggestions.rs stdout ----
diff of stderr:

11         |                             ^^^^ required by this bound in `ConstrainedStruct`
12      help: consider further restricting type parameter `X`
13         |
-       LL | trait InsufficientlyConstrainedGeneric<X=()> where X: std::marker::Copy {
-          |                                              ++++++++++++++++++++++++++
+       LL | trait InsufficientlyConstrainedGeneric<X=()> where X: std::prelude::rust_2021::Copy {
+          |                                              ++++++++++++++++++++++++++++++++++++++
16
17      error: aborting due to previous error
18

```

How can I word the comment better to make it more understandable?

Reply:
I found comment in `visible_parents_map` explaining the order.

https://github.com/rust-lang/rust/blob/2f9c8535bfa70159ccb431174f881b4aa47dc7d7/compiler/rustc_metadata/src/rmeta/decoder/cstore_impl.rs#L296-L299